### PR TITLE
introduce canonical queries, use for normalization and dropck-outlives

### PR DESCRIPTION
**Pull request by nikomatsakis**
This branch adds in the concept of a **canonicalized trait query** and uses it for three specific operations:

- `infcx.at(cause, param_env).normalize(type_foldable)`
    - normalizes all associated types in `type_foldable`
- `tcx.normalize_erasing_regions(param_env, type_foldable)`
    - like normalize, but erases regions first and in the result; this leads to better caching
- `infcx.at(cause, param_env).dropck_outlives(ty)`
    - produces the set of types that must be live when a value of type `ty` is dropped
    - used from dropck but also NLL outlives

This is a kind of "first step" towards a more Chalk-ified approach. It leads to a **big** speedup for NLL, which is basically dominated by the dropck-outlives computation. Here are some timing measurements for the `syn` crate (pre-branch measurements coming soon):

| Commit | NLL disabled | NLL enabled |
| ------- | --- | --- |
| Before my branch | 5.43s | 8.99s |
| After my branch | 5.36s | 7.25s |

(Note that NLL enabled still does *all the work* that NLL disabled does, so this is not really a way to compare the performance of NLL versus the AST-based borrow checker directly.) Since this affects all codepaths, I'd like to do a full perf run before we land anything.

Also, this is not the "final point" for canonicalization etc. I think canonicalization can be made substantially faster, for one thing. But it seems like a reasonable starting point for a branch that's gotten a bit larger than I would have liked.

**Commit convention:** First of all, this entire branch ought to be a "pure refactoring", I believe, not changing anything about external behavior. Second, I've tagged the most important commits with `[VIC]` (very important commit), so you can scan for those. =)

r? @eddyb

### CODE REVIEW

File: src/librustc_traits/normalize_erasing_regions.rs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    tcx: TyCtxt<'_, 'tcx, 'tcx>,
+    goal: ParamEnvAnd<'tcx, Ty<'tcx>>,
+) -> Ty<'tcx> {
+    let ParamEnvAnd { param_env, value } = goal;
+    tcx.sess.perf_stats.normalize_ty_after_erasing_regions.increment();
+    tcx.infer_ctxt().enter(|infcx| {
+        let cause = ObligationCause::dummy();
+        match infcx.at(&cause, param_env).normalize(&value) {
+            Ok(Normalized { value: normalized_value, obligations: _ }) => {
+                //                                   ^^^^^^^^^^^
```
Comment (eddyb) [inline src/librustc_traits/normalize_erasing_regions.rs]: How certain is this? Can you assert that this holds, or would that be too expensive?

Comment (nikomatsakis) [reply src/librustc_traits/normalize_erasing_regions.rs]: Hmm. It's probably worth asserting, or maybe refactoring the types to prove it.

**Comment by nikomatsakis** [comment]
@bors try

I'd like to do a perf run.

**Comment by nikomatsakis** [comment]
cc @rust-lang/release -- I'd like to request a perf run. try build completed, not sure what other steps I need to do.

**Comment by Mark-Simulacrum** [comment]
I'll run it soon, for now just always ping me

**Comment by Mark-Simulacrum** [comment]
Perf started

**Comment by nikomatsakis** [comment]
Here are the results:

[perf comparison](http://perf.rust-lang.org/compare.html?start=27a046e9338fb0455c33b13e8fe28da78212dedc&end=fc2b29dda6f30b073f45f1bd2fab8418ad8c68fd&stat=instructions%3Au)

Also archived here, since I've noticed that perf links tend to go stale:

http://smallcultfollowing.com/pr-48411/rustc%20performance%20data.htm

**Comment by nikomatsakis** [comment]
### Major changes (>10%)

Got worse:

- futures/futures-opt: +10-15%
- regex: 34.4% (patched incremental: compile one)
- regex-opt: 56% (patched incremental: compile one-opt)

Got better

- crates.io-opt: -30.0% on println-opt
- regression-31157, regression-31157-opt: -10% or so
- style-servo-opt: -21.0% (clean-incremental-opt)

### Medium changes (3-10%)

- coercions/coercions-opt: +4%
- piston-image: +6% (clean-incremental, clean-incremental-opt)
- unify-linearly: +9% (clean)

Got better:

- tokio-webpush-simple: -5%
- tokio-webpush-simple-opt: -2% through -5%
- unify-linearly: -4% (dummy fn)

### Small changes (<3%)

- syn: -2%
- crates.io: -2.5% mostly (one change +1%)
- deep-vector: +1%
- htm5ever: +2%
- html5ever-opt: +1%
- hyper: +1%, -2%
- parser, parser-opt: +2%
- piston-image: -1%, +6%
- tuple-stress: +2%

**Comment by nikomatsakis** [comment]
I'll try to take a look at `regex` and `regex-opt`.

**Comment by Mark-Simulacrum** [comment]
Perf started for 251f86541a2fcc1c26534ce4d680fe23f82a8bb7.

**Comment by nikomatsakis** [comment]
@Mark-Simulacrum when I try to compare,I get errors about not having results for  063deba92e44809125a433ca6e6c1ad0993313bf ... am I doing something wrong ? =)

**Comment by nikomatsakis** [comment]
New results available here: [instructions](http://perf.rust-lang.org/compare.html?start=063deba92e44809125a433ca6e6c1ad0993313bf&end=251f86541a2fcc1c26534ce4d680fe23f82a8bb7&stat=instructions%3Au), [wallclock](http://perf.rust-lang.org/compare.html?start=063deba92e44809125a433ca6e6c1ad0993313bf&end=251f86541a2fcc1c26534ce4d680fe23f82a8bb7&stat=wall-time)

Looking at those, and limiting to changes of >5%:

| Test | Improved? | Change | 
| -- | -- | -- |
| futures, futures-opt | 👿  | +11.9% |
| unify-linearly | 👿  | +7.2% |
| unify-linearly-opt | 👿 | 5.6% |
| unused-warnings | 👿 | 5.5% |
| unused-warnings-opt | 👿 | 5% |
| regression-31157-opt | 💚 | about -7% |
| tokio-webpush-simple | 💚 | about -6.6% |
| regression-31157 | 💚 | about -17% |

**Comment by nikomatsakis** [comment]
Interestingly, I can't reproduce the `futures` or `unify-linearly` measurements locally. I tried building with a plain `config.toml` and everything, but the runtimes just come out basically identical. I've not tried with the other tests yet.

**Comment by nikomatsakis** [comment]
I was able to observe the same diffs using the builds from the try server. Interesting. Rebased.

**Comment by nikomatsakis** [comment]
@Mark-Simulacrum if it's easy to do, can you do another perf run, just so that we have up to date data?

**Comment by Mark-Simulacrum** [comment]
Yes, I've already started it. Probably another 30-50 minutes.

**Comment by Mark-Simulacrum** [comment]
http://perf.rust-lang.org/compare.html?start=322d7f7b97f1672bade31e1c43d5753e0e45a1a8&end=94ed417a82daac50123a3e9ee3ed7cd2d308ba94&stat=instructions%3Au

**Comment by nikomatsakis** [comment]
cc @Mark-Simulacrum -- I'd like to see the impact of those latest changes. I found them to be a mild improvement locally, I want to see what the perf server thinks. (Sadly, the 3 or 4 other things I tried all seemed to be regressions.)

**Comment by nikomatsakis** [comment]
Note: the commits I pushed since the try are not important for perf, just new test cases...

**Comment by Mark-Simulacrum** [comment]
Hm, try build seems to have at least partially failed (https://travis-ci.org/rust-lang/rust/jobs/346967602#L9421-L9422), so @bors try

**Comment by nikomatsakis** [comment]
OK, to get tests to pass I had to rebase.

@bors try

**Comment by nikomatsakis** [comment]
cc @Mark-Simulacrum ok -- should be ready for perf comparison now

**Comment by Mark-Simulacrum** [comment]
Perf started.

**Comment by nikomatsakis** [comment]
@Mark-Simulacrum will this be the URL?

http://perf.rust-lang.org/?start=29f5c699b11a6a148f097f82eaa05202f8799bbc&end=f1323669bf496891bbd821112eb5a6f89e15a283&absolute=true&stat=instructions%3Au

**Comment by Mark-Simulacrum** [comment]
That looks right.

**Comment by nikomatsakis** [comment]
@Mark-Simulacrum not working yet though =)

**Comment by Mark-Simulacrum** [comment]
Hm, it looks like we're behind on building master -- I've prioritized the master commit you need, though.

**Comment by nikomatsakis** [comment]
@Mark-Simulacrum oh, weird. I thought I saw results for 29f5c699 in the graphs.

**Comment by nikomatsakis** [comment]
http://perf.rust-lang.org/compare.html?start=bedbad61195d2eae69b43eca49c6d3e2aee8f208&end=29f5c699b11a6a148f097f82eaa05202f8799bbc&stat=instructions:u

**Comment by nikomatsakis** [comment]
perf results available here : http://perf.rust-lang.org/compare.html?start=29f5c699b11a6a148f097f82eaa05202f8799bbc&end=f1323669bf496891bbd821112eb5a6f89e15a283&stat=instructions%3Au

**Comment by nikomatsakis** [comment]
My summary: 

This PR has a big effect on the futures test cases, for which it is an approximately ~20% hit. After a certain amount of profiling, I'm pretty sure that the problem here is that we are not able to get as much benefit from the existing projection cache as we used to get. That cache is tied to the inference context, and so when we move the "dropck-type-elaboration" into a query, it winds up running in another inference context, which doesn't have the cache. This hurts. I believe this will stop hurting when we change our caching scheme to be more global, as I plan to do.

It also regresses piston-image-check by about 10%. I have not investigated this case.

This can also be a substantial win, particularly in some incremental scenarios. For example, crates.io-opt is about 30% faster on the println case, and stylo-server-opt does much better on the "clean incremental" case. I imagine we are able to skip work we couldn't skip before. On the other hand, syn improves by about 7% across the board (and more with incremental) because we getting some caching we didn't get before (that was the primary benchmark I was looking at before).

Other tests are in the <5% range, and are a smatter of small improvements and small regressions.

I'm thinking that the best way forward may be to disable the dropck-elaboration-query **except** when NLL is enabled. I **believe** this would repair the regression for futures, while keeping the benefits for NLL. Moreover, it would let this canonicalization work land, which in turn unlocks progress on traits, which will hopefully let us recoup the perf hit across the board.

**Comment by nikomatsakis** [comment]
cc @rust-lang/wg-traits -- I'm going to have to wrap up for the day. I was thinking that perhaps someone might be interested in helping out with this aspect of this PR:

> I'm thinking that the best way forward may be to disable the dropck-elaboration-query except when NLL is enabled.

It would be perhaps a nice change to familiarize yourself with the overall setup. I can try to lay out what I had in mind. The code in question is the `dropck_outlives` query, [documented here](https://github.com/nikomatsakis/rust/blob/chalkify-canonical-query-mir/src/librustc/traits/query/dropck_outlives.rs#L21-L38). The main way to use this query is to invoke

```rust
infcx.at(&cause, span).dropck_outlives(ty)
```

This indicates that a value of type `ty` is being dropped; it returns back a vector of types which must be live, because a destructor may possible access them. RIght now, it works by internally [creating a canonical query](https://github.com/nikomatsakis/rust/blob/chalkify-canonical-query-mir/src/librustc/traits/query/dropck_outlives.rs#L46) and then invoking a [rustc query also called `dropck_outlives`](https://github.com/nikomatsakis/rust/blob/chalkify-canonical-query-mir/src/librustc/traits/query/dropck_outlives.rs#L48), but this time defined on the global tcx (for more background on rustc queries, see [this chapter of the rustc-guide](https://rust-lang-nursery.github.io/rustc-guide/query.html)). This in turn hands off to [this code in `librustc_traits`](https://github.com/nikomatsakis/rust/blob/chalkify-canonical-query-mir/src/librustc_traits/dropck_outlives.rs#L23-L26).

What I had in mind was to move the code out of `librustc_traits` and (sadly) back into a method `infcx` in `librustc::traits`. Basically, there would be a method like:

```rust
infcx.at(&cause, span).super_dropck_outlives(ty)
```

this would correspond to [this code right here](https://github.com/nikomatsakis/rust/blob/chalkify-canonical-query-mir/src/librustc_traits/dropck_outlives.rs#L30-L136), which we would extract into a helper. That is, while right now it always executes with a fresh inference context, in the new setup it would execute with *either* a fresh context *or* the original one.

Then we would modify [the `infcx.at().dropck_outlives` method](https://github.com/nikomatsakis/rust/blob/chalkify-canonical-query-mir/src/librustc/traits/query/dropck_outlives.rs#L21-L38) so that if the NLL feature is enabled, it uses the query like now, but otherwise skips canonicalization and just invokes `super_dropck_outlives()` with the current infcx.

If anyone is interested, leave a comment or ping me on gitter! Else I'll look at this as soon as I can, perhaps tomorrow. More later, gotta run.

**Comment by aravind-pg** [comment]
Hey @nikomatsakis, I'd be interested in trying to help here, FWIW. Just to break down your outline into slightly more concrete steps so I understand:
- Pull [this code](https://github.com/nikomatsakis/rust/blob/chalkify-canonical-query-mir/src/librustc_traits/dropck_outlives.rs#L30-L136) (i.e. the core logic of the `dropck_outlives` query that runs within the infcx) out into a helper, say `foo` (that takes in an infcx and returns nothing, just mutates it) for now.
- Introduce a new method `infcx.at().super_dropck_outlives(ty)` that just calls `foo`.
- Change `infcx.at().dropck_outlives` so that it if NLL is enabled, then it does what it currently does, otherwise it simply calls `super_dropck_outlives` with the current infcx (without canonicalizing).

If this is correct, my question then is: why have both `foo` and `super_dropck_outlives` when they're both just helpers and the latter just forwards to the former? Why not just call `foo` directly from `dropck_outlives`? Or is the intention that `super_dropck_outlives` itself is the helper (i.e. there's no separate `foo`)?

**Comment by eddyb** [comment]
@aravind-pg I think @nikomatsakis is still referring to `super_dropck_outlives` when he says "helper".

**Comment by nikomatsakis** [comment]
@aravind-pg 

> Or is the intention that super_dropck_outlives itself is the helper (i.e. there's no separate foo)?

That was my intention. (Incidentally, writing up that summary above made me realize how confusing my names are).

To clarify, the flow now is:

- `infcx.at().dropck_outlives()`: canoncalizes, then invokes:
  - `gcx.dropck_outlives()`

the flow would become:

- not in NLL mode:
  - `infcx.at().dropck_outlives()` invokes:
    - `infcx.at().super_dropck_outlives()`
- in NLL mode (i.e., `tcx.sess.nll()` is true)
  - `infcx.at().dropck_outlives()`: canoncalizes, then invokes:
    - `gcx.dropck_outlives()`, which creates a new infcx and invokes:
      - `infcx.at().super_dropck_outlives()`

Though I realize now that an even *easier* way to do this -- albeit at the cost of some NLL performance -- would be to just revert the changes in `librustc_typeck`. Maybe we should start with *that*, just to validate the performance hypothesis.

**Comment by nikomatsakis** [comment]
Or, yet another variant, instead of checking the NLL flag, the librustc_typeck code could just [invoke `super_dropck_outlives` instead](https://github.com/nikomatsakis/rust/blob/8dfd1d2654176c0d4d4df898f24c9c6bea3b6afa/src/librustc_typeck/check/dropck.rs#L302). This way, normal typeck would skip the query, but NLL typeck would still use it.

**Comment by nikomatsakis** [comment]
Thinking more about this:

> Though I realize now that an even easier way to do this -- albeit at the cost of some NLL performance -- would be to just revert the changes in librustc_typeck. Maybe we should start with that, just to validate the performance hypothesis.

This is the obvious step to take first. I'm going to push a quick commit doing this!

For context, the idea here is that we wind up doing the same computation twice: once in the normal type-checker, and once in NLL land. (Eventually, we'll only do it in the NLL type-checker.) 

In this PR as it is, I've consolidated things so that they both invoke the same query. But we can just leave the normal type-checker working as it is. This should *still* help NLL efficiency/correctness, just not quite as much because it won't share work with the ordinary type-checker (yet).

**Comment by aravind-pg** [comment]
OK thanks, makes sense. I agree that reverting `librustc_typeck` is a good first thing to try. Since you are now doing that yourself, it looks like there's no longer anything I need to do? :) If there is do let me know.

**Comment by nikomatsakis** [comment]
Oh bother. It's a touch more complex than I thought, because I pulled one of the helpers into the new `librustc_traits` and made it private. I don't think I'll quite finish this tonight, I can finish up in the morning. 

@aravind-pg -- well, thanks for the offer! =) Tomorrow I'll try to find some more refactoring-like tasks, I've been thinking about it and it seems like these are the best sorts of places to start.

**Comment by nikomatsakis** [comment]
OK, so, yesterday wound up being crazy, but I tested my theory and...it didn't hold up. In fact, I'm starting to get quite confused. I think I've decided to just bisect my way and try to find out where the performance difference comes from. 

I'll track some numbers here as I go.

| Commit | cargo check time on my computer | description |
| --- | --- | --- | 
| 094d8db161 | 1.41, 1.48, 1.7s | just the folding changes, more or less |

**Comment by nikomatsakis** [comment]
Argh, ok, so -- for whatever reason, I am getting pretty unstable numbers on my machine right now. For example, running the merge-base commit, I saw the following:

```bash
lunch-box. time /home/nmatsakis/versioned/rust-1/build/x86_64-unknown-linux-gnu/stage2/bin/rustc --crate-name futures src/lib.rs --crate-type lib --emit=dep-info,metadata -C debuginfo=2 -C metadata=84a8c\
ad17e3c3448 -C extra-filename=-84a8cad17e3c3448 --out-dir /home/nmatsakis/versioned/rustc-perf/collector/benchmarks/futures/target/debug/deps -L dependency=/home/nmatsakis/versioned/rustc-perf/collector/\
benchmarks/futures/target/debug/deps --extern log=/home/nmatsakis/versioned/rustc-perf/collector/benchmarks/futures/target/debug/deps/liblog-ad1e099a7c8f1148.rmeta 2> /dev/null


real    0m1.300s
user    0m1.241s
sys     0m0.036s
lunch-box. lunch-box. time /home/nmatsakis/versioned/rust-1/build/x86_64-unknown-linux-gnu/stage2/bin/rustc --crate-name futures src/lib.rs --crate-type lib --emit=dep-info,metadata -C debuginfo=2 -C met\
adata=84a8cad17e3c3448 -C extra-filename=-84a8cad17e3c3448 --out-dir /home/nmatsakis/versioned/rustc-perf/collector/benchmarks/futures/target/debug/deps -L dependency=/home/nmatsakis/versioned/rustc-perf\
/collector/benchmarks/futures/target/debug/deps --extern log=/home/nmatsakis/versioned/rustc-perf/collector/benchmarks/futures/target/debug/deps/liblog-ad1e099a7c8f1148.rmeta 2> /dev/null


real    0m1.498s
user    0m1.462s
sys     0m0.036s
lunch-box. lunch-box. time /home/nmatsakis/versioned/rust-1/build/x86_64-unknown-linux-gnu/stage2/bin/rustc --crate-name futures src/lib.rs --crate-type lib --emit=dep-info,metadata -C debuginfo=2 -C met\
adata=84a8cad17e3c3448 -C extra-filename=-84a8cad17e3c3448 --out-dir /home/nmatsakis/versioned/rustc-perf/collector/benchmarks/futures/target/debug/deps -L dependency=/home/nmatsakis/versioned/rustc-perf\
/collector/benchmarks/futures/target/debug/deps --extern log=/home/nmatsakis/versioned/rustc-perf/collector/benchmarks/futures/target/debug/deps/liblog-ad1e099a7c8f1148.rmeta 2> /dev/null


real    0m1.298s
user    0m1.258s
sys     0m0.040s
lunch-box. lunch-box. time /home/nmatsakis/versioned/rust-1/build/x86_64-unknown-linux-gnu/stage2/bin/rustc --crate-name futures src/lib.rs --crate-type lib --emit=dep-info,metadata -C debuginfo=2 -C met\
adata=84a8cad17e3c3448 -C extra-filename=-84a8cad17e3c3448 --out-dir /home/nmatsakis/versioned/rustc-perf/collector/benchmarks/futures/target/debug/deps -L dependency=/home/nmatsakis/versioned/rustc-perf\
/collector/benchmarks/futures/target/debug/deps --extern log=/home/nmatsakis/versioned/rustc-perf/collector/benchmarks/futures/target/debug/deps/liblog-ad1e099a7c8f1148.rmeta 2> /dev/null


real    0m1.590s
user    0m1.558s
sys     0m0.032s
lunch-box. lunch-box. time /home/nmatsakis/versioned/rust-1/build/x86_64-unknown-linux-gnu/stage2/bin/rustc --crate-name futures src/lib.rs --crate-type lib --emit=dep-info,metadata -C debuginfo=2 -C met\
adata=84a8cad17e3c3448 -C extra-filename=-84a8cad17e3c3448 --out-dir /home/nmatsakis/versioned/rustc-perf/collector/benchmarks/futures/target/debug/deps -L dependency=/home/nmatsakis/versioned/rustc-perf\
/collector/benchmarks/futures/target/debug/deps --extern log=/home/nmatsakis/versioned/rustc-perf/collector/benchmarks/futures/target/debug/deps/liblog-ad1e099a7c8f1148.rmeta 2> /dev/null


real    0m1.588s
user    0m1.554s
sys     0m0.036s
```

**Comment by nikomatsakis** [comment]
OK, things seem a *bit* more repeatable now. Not sure what was up. Gathering more numbers.

**Comment by nikomatsakis** [comment]
OK, my investigations seem to have finally reached a conclusion. There are two things affecting overall performance here:

- The heavy use of macros in the type-folding code. Specifically, one of the impls that was reverted in 6528ad7ebf. My *guess* would be that it has something to do with the impl for a trait obligation: it used to simply **clone** the `cause`, but now it folds the cause too. This is probably ungood. I can test a targeted revert of *just that impl*.
- The dropck-outlives problem I cited earlier. In my test here, I just reverted the whole change (for NLL too), but I can try a targeted revert there as well.

**Comment by nikomatsakis** [comment]
.....and, at the end of the day, this appears to be the crucial diff:

```diff
modified   src/librustc/traits/structural_impls.rs
@@ -11,7 +11,7 @@
 use traits;
 use traits::project::Normalized;
 use ty::{self, Lift, TyCtxt};
-use ty::fold::{TypeFoldable};
+use ty::fold::{TypeFoldable, TypeFolder, TypeVisitor};

 use std::fmt;
 use std::rc::Rc;
@@ -348,10 +348,20 @@ impl<'a, 'tcx> Lift<'tcx> for traits::Vtable<'a, ()> {
 ///////////////////////////////////////////////////////////////////////////
 // TypeFoldable implementations.

-BraceStructTypeFoldableImpl! {
-    impl<'tcx, O> TypeFoldable<'tcx> for traits::Obligation<'tcx, O> {
-        cause, recursion_depth, predicate, param_env
-    } where O: TypeFoldable<'tcx>
+impl<'tcx, O: TypeFoldable<'tcx>> TypeFoldable<'tcx> for traits::Obligation<'tcx, O>
+{
+    fn super_fold_with<'gcx: 'tcx, F: TypeFolder<'gcx, 'tcx>>(&self, folder: &mut F) -> Self {
+        traits::Obligation {
+            cause: self.cause.clone(),
+            recursion_depth: self.recursion_depth,
+            predicate: self.predicate.fold_with(folder),
+            param_env: self.param_env.fold_with(folder),
+        }
+    }
+
+    fn super_visit_with<V: TypeVisitor<'tcx>>(&self, visitor: &mut V) -> bool {
+        self.predicate.visit_with(visitor)
+    }
 }

 BraceStructTypeFoldableImpl! {
```

I'm ... happy I guess but sad it took me so long to find that. =)

**Comment by nikomatsakis** [comment]
Weird but true: that final commit makes things a lot slower. I have no idea why ...

File: src/librustc/traits/structural_impls.rs [hunk 2]
```diff
@@ -348,19 +348,15 @@ impl<'a, 'tcx> Lift<'tcx> for traits::Vtable<'a, ()> {
...
-        }
-    }
+BraceStructTypeFoldableImpl! {
+    impl<'tcx, O> TypeFoldable<'tcx> for traits::Obligation<'tcx, O> {
+        cause, recursion_depth, predicate, param_env
+    } where O: TypeFoldable<'tcx>
+}
 
-    fn super_visit_with<V: TypeVisitor<'tcx>>(&self, visitor: &mut V) -> bool {
-        self.predicate.visit_with(visitor)
```
Comment (nikomatsakis) [inline src/librustc/traits/structural_impls.rs]: oh, I guess it's the fact that `visit_with` is broken here

**Comment by nikomatsakis** [comment]
@Mark-Simulacrum can you do a perf run by any chance? [try run succeeded](https://github.com/rust-lang/rust/pull/48411#issuecomment-370160305)

**Comment by nikomatsakis** [comment]
@bors try

Another day, another merge conflict.

**Comment by Mark-Simulacrum** [comment]
Started perf for 030b91ae711216469a85c46508c4010c2f43e1a0.

**Comment by nikomatsakis** [comment]
@Mark-Simulacrum thanks! I suppose this will be the URL (created from the compare page):

http://perf.rust-lang.org/compare.html?start=c9b86a91786ca54e898b54f30a6e11546b8e8026&end=030b91ae711216469a85c46508c4010c2f43e1a0&stat=instructions%3Au

(I don't see any data for 030b91ae711216469a85c46508c4010c2f43e1a0 yet though)

**Comment by Mark-Simulacrum** [comment]
Not sure what went wrong -- I've restarted the build and will try to keep an eye on it.

**Comment by nikomatsakis** [comment]
@Mark-Simulacrum ok -- I just had to rebase, so I did another try as well.

**Comment by nikomatsakis** [comment]
cc @rust-lang/compiler but especially @eddyb  : 

What are your thoughts on the [performance details I summarized here](https://github.com/nikomatsakis/rustc-guide/blob/c450f94c0f3761d68241307fd9c0cb29349ebe2e/src/traits-lowering-rules.md#implied-bounds). I still have a few nits from @eddyb to correct, but presuming that I do that, should I merge as is? Or is it worth putting more effort into tracking down the causes of `coercions-check` etc?

**Comment by Mark-Simulacrum** [comment]
I've queued 61b468e3858ee052d443ec92fe988c8050e56101 as well for a perf build -- though I don't expect any major changes.

----

Regarding the regressions, it looks like we aren't actually regressing by that much -- the 3-5% changes are timewise <0.1 seconds in most or all cases so probably not worth being concerned about. I'd merge as-is (perhaps pending nits by @eddyb, but that's somewhat unrelated).

**Comment by nikomatsakis** [comment]
I did some more analysis, looking closely at cargo-check. The good news (sort of) is that *now* the bottleneck appears to be the canonicalization code that this PR introduces; in the branch we spend about 6% of total time doing `check_safety_of_destructor_if_necessary` (vs 0% on master), which is *basically* all spent canonicalizing the queries and instantiating the responses (see detailed accounting below). The reason this strikes me as good news is that 

- it means we've eliminated all the "accidental bottlenecks";
- computing the *results* doens't seem to be expensive, so the "cache population" problem I was worried about doesn't seem to be a big factor;
- optimizing the canonicalization and instantiation process seems tractable

That said, I don't know how to do the final bullet yet, still digging in a bit. But it might be something where we can land this PR and focus in on that final bullet. I should probably check a few other tests, but given that the regression range is similar, I am hopeful that they will still be showing 

<details>
<summary>

detailed look at where `check_safety_of_destructors` spends its time

</summary>

```bash
> perf focus '{check_safety_of_destructor_if_necessary}' --tree-callees --min-percent 5 --relative
Matcher    : {check_safety_of_destructor_if_necessary}
Matches    : 499
Not Matches: 7593
Percentage : 100
Time       : 0.50s at 1000 Hz

Tree
| matched `{check_safety_of_destructor_if_necessary}` (100% total, 3% self)
: | rustc::traits::query::dropck_outlives::<impl rustc::infer::at::At<'cx, 'gcx, 'tcx>>::dropck_outlives (91% total, 2% self)
: : | rustc::infer::canonical::<impl rustc::infer::InferCtxt<'cx, 'gcx, 'tcx>>::instantiate_query_result (47% total, 2% self)
: : : | rustc::infer::canonical::<impl rustc::infer::InferCtxt<'cx, 'gcx, 'tcx>>::unify_canonical_vars (28% total, 1% self)
: : : : | rustc::infer::InferCtxt::commit_if_ok (13% total, 2% self)
: : : : : | rustc::infer::InferCtxt::commit_from (7% total, 0% self)
: : : : | rustc::infer::InferCtxt::start_snapshot (6% total, 4% self)
: : : : | rustc::infer::InferCtxt::commit_from (5% total, 1% self)
: : : | <rustc::infer::canonical::Canonical<'tcx, V>>::substitute (8% total, 1% self)
: : | rustc::infer::canonical::Canonicalizer::canonicalize (30% total, 1% self)
: : : | <rustc::infer::canonical::Canonicalizer<'cx, 'gcx, 'tcx> as rustc::ty::fold::TypeFolder<'gcx, 'tcx>>::fold_ty (23% total, 0% self)
: : : : | rustc::ty::structural_impls::<impl rustc::ty::fold::TypeFoldable<'tcx> for &'tcx rustc::ty::TyS<'tcx>>::super_fold_with (21% total, 1% self)
: : : : : | <rustc::infer::canonical::Canonicalizer<'cx, 'gcx, 'tcx> as rustc::ty::fold::TypeFolder<'gcx, 'tcx>>::fold_region (11% total, 0% self)
: : : : : : | <std::collections::hash::map::HashMap<K, V, S>>::entry (5% total, 1% self)
```

</details>

**Comment by nikomatsakis** [comment]
futures-check seems to have a somewhat different profile. It is still spending time in that general area, but it's doing it differently. Approximately 50% of the time is spent on computation, 35% on canonicalization, and 15% processing the results (e.g., `projection_declared_bounds`). This is mildly surprising to me. I'll try to dig in a bit more and see if what it suggests.

**Comment by nikomatsakis** [comment]
Rebased and added a few simple micro-optimizations. These made a 2 or 3% difference in futures-check locally.

**Comment by nikomatsakis** [comment]
cc @Mark-Simulacrum ok it's done

**Comment by Mark-Simulacrum** [comment]
Perf started. Will, probably, be available in ~2 hours at http://perf.rust-lang.org/compare.html?start=5430c0c5c0fbdfb8e89358a187d2f9a8d4b796d4&end=d0868c1d6f0efac98e5312316dbf906ffa2d1b49&stat=instructions%3Au.

**Comment by nikomatsakis** [comment]
@Mark-Simulacrum 

hero

**Comment by nikomatsakis** [comment]
OK, this [latest run](http://perf.rust-lang.org/compare.html?start=5430c0c5c0fbdfb8e89358a187d2f9a8d4b796d4&end=d0868c1d6f0efac98e5312316dbf906ffa2d1b49&stat=instructions%3Au) is looking great, I'm going to declare victory on the perf front. I'll rebase for the 222nd time, address @eddyb's nits, and then land this. 🎉

**Comment by sgrif** [comment]
Holy shit that perf run looks awfully purty

**Comment by nikomatsakis** [comment]
OK, rebased. *Again.*

**Comment by nikomatsakis** [comment]
@eddyb 

Earlier you wrote this:

> Can we get rid of trans_apply_param_substs now?

At this point, `trans_apply_param_substs` is just a combination of "apply the substitute, normalize the result". But that seems like a useful combination. Is it the name that's bothering you? Maybe we should rename to `subst_and_normalize_erasing_regions`? 

There are also two variants, one that takes an environment and one that doesn't. The latter is hardly used and I'd probably just remove it.

**Comment by nikomatsakis** [comment]
OK, I think I addressed both of @eddyb's nits here. Hopefully they agree!

**Comment by alexcrichton** [comment]
Failed to document, how anticlimactic :(

```
[01:52:13] [m[m[32m[1m Documenting[m syntax_ext v0.0.0 (file:///checkout/src/libsyntax_ext)
[01:52:53] error: character literal may only contain one codepoint: '?
[01:52:53]  --> <stdin>:1:9
[01:52:53]   |
[01:52:53] 1 | ?0: Foo<'?1, ?2>
[01:52:53]   |         ^^
[01:52:53] 
[01:52:54] [m[m[31m[1merror:[m Could not document `rustc`.
[01:52:54] 
```

**Comment by nikomatsakis** [comment]
@alexcrichton *grrrrr* =) wait, I'm confused...

Did we change our doc requirements or something here? (And I saw a green check before...)

**Comment by nikomatsakis** [comment]
But also, that string does not appear in `syntax_ext`

File: src/librustc/infer/canonical.rs [hunk 3]
```diff
@@ -0,0 +1,988 @@
...
+    /// instantiates the result so it can be used, plugging in the
+    /// values from the canonical query. (Note that the result may
+    /// have been ambiguous; you should check the certainty level of
+    /// the query before applying this function.)
+    ///
+    /// It's easiest to explain what is happening here by
+    /// example. Imagine we start out with the query `?A: Foo<'static,
+    /// ?B>`. We would canonicalize that by introducing two variables:
+    ///
+    ///     ?0: Foo<'?1, ?2>
```
Comment (bjorn3) [inline src/librustc/infer/canonical.rs]: @nikomatsakis 
> But also, that string does not appear in syntax_ext

It appears here

Comment (nikomatsakis) [reply src/librustc/infer/canonical.rs]: Yes, I now, thanks =) I just thought we didn't test examples in librustc (imo, we still shouldn't, but maybe we changed that?)

Comment (bjorn3) [reply src/librustc/infer/canonical.rs]: Didn't notice your commit before I commented :)

Comment (nikomatsakis) [reply src/librustc/infer/canonical.rs]: no worries =) in any case, I think maybe we still try to *parse* even if we don't run tests...

Comment (nikomatsakis) [reply src/librustc/infer/canonical.rs]: (for some weird reason `./x.py test` is also giving me other strange, doc-test-related errors locally...)

Comment (RalfJung) [reply src/librustc/infer/canonical.rs]: Couldn't you have added `ignore`? I think I've seen this in librustc recently.

Comment (SimonSapin) [reply src/librustc/infer/canonical.rs]: > other strange, doc-test-related errors

If it’s `E0003` in the error index, try `rm build/tmp/extended-error-metadata/x86_64-unknown-linux-gnu/librustc_const_eval.json`

**Comment by nikomatsakis** [comment]
@bors 

> 🔒 Merge conflict

Curse you @bors!

**Comment by aravind-pg** [comment]
Why not give it a higher priority?

**Comment by nikomatsakis** [comment]
We could, I suppose, given that we're trying to build on it. Note though that PRs get priority by their age, so it's already *reasonably* high priority -- though I see there are a few PRs in front of it in the queue.

**Comment by nikomatsakis** [comment]
Looking at the travis log, the only error I see was this:

```
[01:51:19] env_logger/struct.Logger.html:101: broken link fragment `#tymethod.flush` pointing to `log/trait.Log.html`
[01:51:22] thread 'main' panicked at 'found some broken links', tools/linkchecker/main.rs:49:9
```

**Comment by nikomatsakis** [comment]
Just in case, did a rebase, trying a local build.

**Comment by kennytm** [comment]
Broken link in `env_logger` 😂 

```
[01:51:19] env_logger/struct.Logger.html:101: broken link fragment `#tymethod.flush` pointing to `log/trait.Log.html`
[01:51:22] thread 'main' panicked at 'found some broken links', tools/linkchecker/main.rs:49:9
[01:51:22] note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

**Comment by nikomatsakis** [comment]
@bors r=eddyb 

Hopefully that fixes it!

**Comment by nikomatsakis** [comment]
cleaned up history -- I bumped the log dependency to 0.4 though in `librustc_traits`, so hopefully that fixes it

**Pull request merged**